spiketoolkit/postprocessing/postprocessing_tools.py

```diff
@@ -164,27 +254,65 @@ def get_unit_waveforms(recording, sorting, unit_ids=None, grouping_property=None
+                    spike_index_list.append(indices)
     else:
+        if max_channels_per_waveforms is None:
+            max_channels_per_waveforms = recording.get_num_channels()
+
         for i, unit_id in enumerate(unit_ids):
             if unit_id not in sorting.get_unit_ids():
                 raise Exception("unit_ids is not in valid")
 
             fs = recording.get_sampling_frequency()
             n_pad = [int(ms_before * fs / 1000), int(ms_after * fs / 1000)]
 
+            if max_spikes_per_unit is None:
+                max_spikes = len(sorting.get_unit_spike_train(unit_id))
+            else:
+                max_spikes = max_spikes_per_unit
+
             if verbose:
                 print('Waveform ' + str(i + 1) + '/' + str(len(unit_ids)))
-            waveforms, indices = _get_random_spike_waveforms(recording=recording,
-                                                             sorting=sorting,
-                                                             unit=unit_id,
-                                                             max_spikes_per_unit=max_spikes_per_unit,
-                                                             snippet_len=n_pad,
-                                                             channel_ids=channel_ids,
-                                                             seed=seed)
-            waveforms = waveforms.swapaxes(0, 2)
-            waveforms = waveforms.swapaxes(1, 2)
-            waveforms = waveforms.astype(dtype)
+            wf, indices = _get_random_spike_waveforms(recording=recording,
+                                                      sorting=sorting,
+                                                      unit=unit_id,
+                                                      max_spikes_per_unit=max_spikes,
+                                                      snippet_len=n_pad,
+                                                      channel_ids=channel_ids,
+                                                      seed=seed)
+            wf = wf.swapaxes(0, 2)
+            wf = wf.swapaxes(1, 2)
+            wf = wf.astype(dtype)
+
+            if max_channels_per_waveforms < recording.get_num_channels():
+                template = np.mean(wf, axis=0)
+                # select based on adjacency
+                if 'location' in recording.get_shared_channel_property_names():
+                    max_channel_idx = np.unravel_index(np.argmax(np.abs(template)),
+                                                       template.shape)[0]
+                    locs = recording.get_channel_locations()
+                    loc_max = locs[recording.get_channel_ids().index(max_channel_idx)]
+                    distances = np.array([np.linalg.norm(l - loc_max) for l in locs])
+                    max_channel_idxs = np.argsort(distances)[:max_channels_per_waveforms]
+                    wf = wf[:, max_channel_idxs]
+                else:  # select based on amplitude
```

Comment:
All that stuff is a copy paste from previous code, it should capitalize in a function.

Reply:
ok

Reply:
made a new function